garak/detectors/fileformats.py

```diff
@@ -0,0 +1,86 @@
+# SPDX-FileCopyrightText: Portions Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""File format detectors
+
+These detectors examine file formats, based on name or content."""
+
+import pickletools
+from typing import Iterable
+
+import huggingface_hub
+import tqdm
+
+# import huggingface_hub.utils
+
+from garak.attempt import Attempt
+import garak._config
+from garak.detectors.base import Detector
+
+
+class PossiblePickleName(Detector):
+    """Guess if filenames look like pickles
+
+    Pickle files have multiple and complex vulnerabilities, and should be
+    avoided. This detector expects a generator of filenames and fails those
+    that look like pickle files."""
+
+    def detect(self, attempt: Attempt) -> Iterable[float]:
+        for filename in attempt.outputs:
+            if filename is None:
+                continue
+            elif filename.lower().endswith(".pkl"):
+                yield 1.0
+            elif filename.lower().endswith(".pickle"):
+                yield 1.0
+            else:
+                yield 0.0
+
+
+class HF_IsPickled(Detector):
+    """
+    Check whether a file is pickled.
+    """
+
+    def detect(self, attempt: Attempt) -> Iterable[float]:
+        if "model_name" not in dir(garak._config.plugins):
+            return [0.0] * len(attempt.outputs)
+        hf_repo_name = garak._config.plugins.model_name
+        for repo_filename in tqdm.tqdm(
+            attempt.outputs, leave=False, desc=f"Checking for pickles in {hf_repo_name}"
+        ):
+            if repo_filename is None:
+                continue
+            else:
+                try:
+                    # huggingface_hub.utils.disable_progress_bars()
+                    local_filename = huggingface_hub.hf_hub_download(
+                        hf_repo_name, repo_filename
+                    )
+                    # huggingface_hub.utils.enable_progress_bars()
+                    with open(local_filename, "rb") as f:
+                        pkl_data = pickletools.genops(f)
+
+                        # If the file is not a pickle, we may hit a UnicodeDecode error or ValueError when running over genops
+                        *_, last = pkl_data
```

Comment:
Am I reading this right as the `detector` is downloading files?  This seems like the wrong placement of responsibility based on existing plugin types.

I currently expect a `probe` to gather information and pass it to a `detector`. On the other hand this also introduces the concept that gathered data could be very `large` and difficult to retain as evidence for results. In this case we might need to, at at a minimum, also generate *version* or *fingerprint* details for these larger pieces of data that can be referenced in a report.

Reply:
This is the method for single-file hugging face cache access. If the model's used for generation, this is a noop.

It's def infeasible to pass file contents in attempts and I don't think a summary is an efficient way forward. We could do the local fs cache lookup on the probe side, though, iff this preserves repo filenames.

Reply:
"This is cached access" makes it reasonable to say this instance is really accessing no not obtaining data, that works for me on clarifying the "responsibility for download" concern.

Based on that I would like to suggest some refactoring though. Any test would currently need to intercept `huggingface_hub.hf_hub_download` to provide the object detection actually acts upon, where as if the data was provided as an attributed of the `Attempt` being evaluated the test could just instantiate an Attempt locally without coupling to `huggingface` retrieval. I am not suggesting this PR should adjust `Attempt` just yet however we could extract the actual detection to be testable.

To that end it may be reasonable to shift the actual `detection` into a separate method maybe `IsPickled(filename: str)` or `IsPickled(filename: []byte)`that is called by `HF_IsPickled()`. I am thinking that a more isolated `detection` may be more consumable for other use cases, such as if at some point we supported `text[in] -> binary[out]` modality and wanted to have a detector that could test if a model can is returning binary data that is a `pickled` object.

Reply:
This makes a tonne of sense. Moved responsibility for getting file data away from `detector` to `probe`; this keeps the detector flexible wrt. where the file came from, i.e. one file checker can work for any generator, and code that looks inside the file is decoupled from code that gets the file. Under this pattern, `detectors.fileformats.FileIsPickled.detect()` is pretty much a method that only looks at that item.